Sources/SwiftCompilerPluginMessageHandling/Macros.swift

```diff
@@ -43,11 +44,26 @@ extension CompilerPluginMessageHandler {
         throw MacroExpansionError.freestandingMacroSyntaxIsNotMacro
       }
       guard let macroDefinition = resolveMacro(macro) else {
-        throw MacroExpansionError.macroTypeNotFound
+        throw MacroExpansionError.macroTypeNotFound(macro)
+      }
+
+      let macroRole: MacroRole
+      if let pluginMacroRole {
+        switch pluginMacroRole {
+        case .expression: macroRole = .expression
+        case .declaration: macroRole = .declaration
+        case .codeItem: macroRole = .codeItem
+
+        case .accessor, .conformance, .member, .memberAttribute, .peer:
+          throw MacroExpansionError.invalidMacroRole(pluginMacroRole)
+        }
```

Comment:
(Not critical at all)
Could you use `MacroRole.init(messageMacroRole: PluginMessage.MacroRole)` here?
Since `expandFreestandingMacro` diagnoses unmatched macro role, diagnosing invalid is not necessary I think.

Reply:
But if we do diagnose invalid role here, we should do it in `expandAttachedMacro()` too

Reply:
I'd rather be defensive here, because we're on one side of an IPC barrier

Reply:
Nevermind, I see what you mean. It'll be checked and produce an error down in `expandFreestandingMacro`, so we don't need the redundant check here.

Reply:
Cleanup is at https://github.com/apple/swift-syntax/pull/1762